cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -80,6 +80,134 @@ const FunctionDoc list_value_length_doc{
      "Null values emit a null in the output."),
     {"lists"}};
 
+template <typename Type, typename IndexType>
+Status ListElementArray(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  using ListArrayType = typename TypeTraits<Type>::ArrayType;
+  using IndexScalarType = typename TypeTraits<IndexType>::ScalarType;
+  const auto& index_scalar = batch[1].scalar_as<IndexScalarType>();
+  if (ARROW_PREDICT_FALSE(!index_scalar.is_valid)) {
+    return Status::Invalid("Index must not be null");
+  }
+  ListArrayType list_array(batch[0].array());
+  auto index = index_scalar.value;
+  if (ARROW_PREDICT_FALSE(index < 0)) {
+    return Status::Invalid("Index ", index,
+                           " is out of bounds: should be greater than or equal to 0");
+  }
+  std::unique_ptr<ArrayBuilder> builder;
+  RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), list_array.value_type(), &builder));
+  RETURN_NOT_OK(builder->Reserve(list_array.length()));
+  for (int i = 0; i < list_array.length(); ++i) {
+    if (list_array.IsNull(i)) {
+      RETURN_NOT_OK(builder->AppendNull());
+      continue;
+    }
+    std::shared_ptr<arrow::Array> value_array = list_array.value_slice(i);
+    auto len = value_array->length();
+    if (ARROW_PREDICT_FALSE(index >= static_cast<typename IndexType::c_type>(len))) {
+      return Status::Invalid("Index ", index, " is out of bounds: should be in [0, ", len,
+                             ")");
+    }
+    RETURN_NOT_OK(builder->AppendArraySlice(*value_array->data(), index, 1));
+  }
+  ARROW_ASSIGN_OR_RAISE(auto result, builder->Finish());
+  out->value = result->data();
+  return Status::OK();
+}
+
+template <typename, typename IndexType>
+struct ListElementScalar {
```

Comment:
Why is there an anonymous template parameter?

Reply:
GenerateInteger unconditionally passes in at least one other template parameter. If we really want to get rid of it, we could make it `typename ListScalarType` and call `GenerateInteger<ListElementScalar, BaseListScalar>`.